### PR TITLE
Route custom process groups through GDP and latent MoE checkpoints

### LINKED ISSUE #6501: [feat] GTP+DCP
GTP+DCP(Distributed Checkpoint) has been supported in the GTP init MR #4967 

This issue will keep tracking following-up bug-fixes / perf optimizations for GTP+DCP related stuffs.
- fix GTP+DCP for GDP moudle (#6503 )
- fix GTP+DCP when parallel_state is not initialized (#6775 )
- fix the case when some params is sharded across GTP while been duplicated across TP group (e.g., the latent gemm) (#6781 )
- GTP+DCP+fp8-param save->load is not bitwise matching 
        - MXFP8 (bytes + scale_A) → dequantize → BF16 → save → load → quantize → MXFP8 (bytes + scale_B)
               - Where scale_A (from fp32 master w) ≠ scale_B
               - quantize(dequantize(mfxp8) is a lossy round-trip, not identical
        - solution: during loading: cast from fp32 main params into mxfp8 params instead (related MR: #6666 )
- Fix GTP checkpoint load failing across different GTP degrees (#7293)

### PR DESCRIPTION
**Pull request by yashaswikarnati**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?

Route explicit model-parallel process groups through GDP and latent-MoE distributed-checkpoint metadata without changing duplicated TELinear execution semantics.

MIMO models can own process groups that differ from the globally initialized MPU topology. This change:

- uses the GDP mixer's explicit TP and DP+CP groups for RMSNorm, base parameters, Conv1d, and child-module checkpoint wrappers;
- passes the owning TP group to latent-MoE projection linears;
- retains that group on duplicated `TELinear` only for checkpoint metadata, while Transformer Engine still receives `tp_group=None`, `tp_size=1`, and `parallel_mode=None`;
- preserves the current GTP `in_proj` checkpoint split and resharding path.

This is intentionally separate from LayerWise optimizer changes, MTP, optimizer policy, recipes, and padding-only checkpoint work.

## Issue tracking

Linked issue: N/A

### Pre-checks

- [x] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [x] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

Validation:

- `CHECK_ONLY=true uv run --no-sync bash tools/autoformat.sh` (Black, isort, Pylint 10.00/10, and Ruff passed; mypy was unavailable and is non-blocking in the script)
- `python3 -m compileall -q` on all six changed Python files
- CW-DFW environment job `16387980`: 1 H100; locked test environment preflight passed with PyTorch 2.13 nv26.06, Transformer Engine 2.16, pytest 9.1.1, and pytest-mock
- CW-DFW job `16388633`: 1 node, 8 H100s, 8 ranks, current `26.06-py3` development image; all ranks passed the 6 focused GDP custom-group/TP-reshard, latent-MoE construction, and duplicated-TELinear execution/checkpoint tests

### CODE REVIEW

**Review by Phlip79 — APPROVED** [review]

File: megatron/core/ssm/gated_delta_product.py [hunk 1]
```diff
@@ -127,9 +127,15 @@ class ExtendedRMSNorm(RMSNormGated):
...
+        metadata = ensure_metadata_has_dp_cp_group(metadata)
         state_dict = self.state_dict(prefix="", keep_vars=True)
         return make_sharded_tensors_for_checkpoint(
-            state_dict, prefix, {"weight": 0}, sharded_offsets
+            state_dict,
+            prefix,
+            {"weight": 0},
+            sharded_offsets,
+            tp_group=self.tp_group,
+            dp_cp_group=metadata["dp_cp_group"],
```
Comment (Phlip79) [inline megatron/core/ssm/gated_delta_product.py:138]: Why do we not just use `dp_cp_group` directly?

**Review by Phlip79 — APPROVED** [review]

File: megatron/core/extensions/transformer_engine.py [hunk 2]
```diff
@@ -1227,7 +1227,8 @@ def __init__(
             ), "Must have at least TE version 2.3 or higher to use symmetric memory all reduce"
             extra_kwargs["symmetric_ar_type"] = symmetric_ar_type
         if parallel_mode == "duplicated":
-            assert tp_group is None, "duplicated linear should not have tp_group set"
```
Comment (fanshiqing) [inline megatron/core/extensions/transformer_engine.py:1230]: the bottom line: we don't require tp_group for `duplicated` mode, 
pls try to figure out a better fix if possible

**Review by fanshiqing — CHANGES_REQUESTED** [review]

File: megatron/core/ssm/gated_delta_product.py (hunk 1 above)
Comment (yashaswikarnati) [reply megatron/core/ssm/gated_delta_product.py:138]: This is just following the existing convention as other layers like Mamba, moe, TE linear layers. I think the above helper function is added ensure_metadata_has_dp_cp_group , so if hte group does not exist, it falls back to paralel state for backward compatibility

File: megatron/core/tensor_parallel/generalized_tensor_parallelism.py [hunk 3]
```diff
@@ -2790,16 +2789,15 @@ def make_sharded_tensors_for_checkpoint_with_gtp_remat(
...
-        # GTP-sharded tensor: delegate to the GTP-aware single-tensor helper — it layers the
-        # axis-0 GTP split onto TP, elects the writer over the gtp_remat-excluded DP group, and sets
-        # allow_shape_mismatch for alignment padding. (tp_axis None → 0; tp_size 1 when no TP.)
+        # Only tensors present in the axis map are also sharded across TP.
         tp_axis = tensor_parallel_layers_axis_map.get(layer_name, None)
         sharded_state_dict[layer_key] = make_tp_sharded_tensor_for_checkpoint(
             tensor,
             layer_key,
             tp_axis=tp_axis if tp_axis is not None else 0,
+            replica_id=None if tp_axis is not None else (0, tp_rank, dp_replica_rank),
```
Comment (fanshiqing) [inline megatron/core/tensor_parallel/generalized_tensor_parallelism.py:2798]: Thanks for fixing this!

That's real bug after supporting `--gtp-remat-opt-in-modules moe_latent_proj` 2 weeks ago where the latent weights are DUPLICATED across TP group while are SHARDED across GTP groups!

File: megatron/core/transformer/moe/moe_layer.py [hunk 4]
```diff
@@ -276,6 +276,9 @@ def __init__(
                 if "moe_latent_proj" in self.config.gtp_remat_opt_in_modules
                 else None
             )
+            linear_gtp_kwargs = {"gtp_remat_group": gtp_remat_group}
+            if linear_cls is TELinear:
```
Comment (fanshiqing) [inline megatron/core/transformer/moe/moe_layer.py:280]: you should also put `linear_gtp_kwargs = {"gtp_remat_group": gtp_remat_group}` under this gurad, where when linear_cls is `InferenceLinear`, it's init also don't accept `gtp_remat_group` at least for now

Comment (yashaswikarnati) [reply megatron/core/transformer/moe/moe_layer.py:280]: done, InferenceLinear is not in the scope for this PR for nwo

**Review by deepakn94 — APPROVED** [review]

**Comment by yashaswikarnati** [comment]
/ok to test [1633cb5](https://github.com/NVIDIA/Megatron-LM/pull/6781/commits/1633cb52e5c14555767d810b7bf6122f22121fdf)

**Review by fanshiqing — APPROVED** [review]
